libtenzir/src/detail/string.cpp

```diff
@@ -18,6 +18,235 @@
 namespace tenzir {
 namespace detail {
 
+auto quoting_escaping_policy::is_inside_of_quotes(
+  std::string_view text, size_t idx) const noexcept -> bool {
+  if (idx > text.size()) {
+    return false;
+  }
+  auto start = size_t{0};
+  while (start < text.size()) {
+    auto open = find_opening_quote(text, start);
+    if (open > idx) {
+      return false;
+    }
+    auto close = find_closing_quote(text, open);
+    if (close == text.npos) {
+      return false;
+    }
+    if (close > idx) {
+      return true;
+    }
+    start = close + 1;
+  }
+  return false;
+}
+
+auto quoting_escaping_policy::find_opening_quote(
+  std::string_view text, size_t start) const -> std::string_view::size_type {
+  auto active_escape = false;
+  for (; start < text.size(); ++start) {
+    const auto character = text[start];
+    const auto is_quote = is_quote_character(character);
+    const auto is_backslash = character == '\\';
+    if (is_quote and not active_escape) {
+      return start;
+    } else if (is_backslash and not active_escape) {
+      active_escape = backslashes_escape;
+      continue;
+    }
+    active_escape = false;
+  }
+  return text.npos;
+}
+
+auto quoting_escaping_policy::find_closing_quote(
+  std::string_view text, size_t opening) const -> std::string_view::size_type {
+  auto active_escape = false;
+  TENZIR_ASSERT(is_quote_character(text[opening]));
+  for (size_t i = opening + 1; i < text.size(); ++i) {
+    const auto character = text[i];
+    const auto is_backslash = character == '\\';
+    if (not active_escape and character == text[opening]) {
+      if (doubled_quotes_escape and i < text.size() - 1
+          and text[i + 1] == character) {
+        ++i;
+        continue;
+      }
+      return i;
+    } else if (is_backslash and not active_escape) {
+      active_escape = backslashes_escape;
+      continue;
+    }
+    active_escape = false;
+  }
+  return text.npos;
+}
+
+auto quoting_escaping_policy::find_first_of_not_in_quotes(
+  std::string_view text, std::string_view targets,
+  size_t start) const -> std::string_view::size_type {
+  auto quote_start = text.npos;
+  auto active_escape = false;
+  for (size_t i = start; i < text.size(); ++i) {
+    const auto character = text[i];
+    const auto maybe_closing = quote_start != text.npos
+                               and character == text[quote_start]
+                               and not active_escape;
+    if (maybe_closing) {
+      if (doubled_quotes_escape and i < text.size() - 1
+          and text[i + 1] == character) {
+        ++i;
+      } else {
+        quote_start = text.npos;
+      }
+      continue;
+    }
+    const bool is_quote = is_quote_character(character);
+    if (is_quote and not active_escape) {
+      quote_start = i;
```

Comment:
Aren't we potentially overriding an existing `quote_start` here if we have multiple quote characters? Ie. for a string like `"foo'bar"?